slice/slice.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package slice provides utility methods for common operations on slices.
+package slice
```

Comment:
From @thockin https://github.com/kubernetes/kubernetes/pull/101194#issuecomment-821455532:
> I would be OK with that idea, too, but I would probably ask for the API to change. e.g.
> `slices.String(variable).Copy()`

I'm quite in favor of creating a:
```
package strings
type Slice []string

func NewSlice(strs ...string) Slice { ... }
```

I think that'd help a lot more with the chaining mentioned in `SortStrings`. I'd also remove the modifier from `ContainString` and `RemoveString` and replace with maybe `Filter`/`Map` though these would require copying (if you're creative you can use a channel). What do you think?

Reply:
Hum, I'm gonna try then. I'll close the other PR and move all the effort here, adding Tim's comments.

About being creative and using channels, I can do for sure, will just try to start with something simpler now and evolve in this PR (I'm bad at channels :P )

Reply:
Is the discussion about `slice.String(s).Whatever()` vs `strings.Slice(s).Whatever()` ?

If we had generics, I think we'd obviously lean towards `slices.Slice[string]` as a pattern wouldn't we?  On the other hand, `strings` exists as a package already...

Reply:
> Is the discussion about `slice.String(s).Whatever()` vs `strings.Slice(s).Whatever()` ?

Yeah, not sure it matters so much? Maybe I've seen a discussion about it somewhere.

`slices.Slice` is a little "smurfy", but yeah, we would probably do `Slice[string]` (or whatever the syntax is). But that's a good point, even if we wanted to generate slices for arbitrary types, we would probably all have them in a slice package, and then just generate `strings.go`, `ints.go` etc.

Reply:
Well, this subject being 'decided' I can move back to slice package and have some strings.go again there.

Now, about the "modifiers" function I'm having some hard time figuring out why is this necessary. I've taken a look into Kubernetes code, and every part that uses slices.ContainsString or slices.RemoveString does not passes the modifier function, but instead 'nil'.

So I'm thinking if in this case is worth to migrate also with the modifier (or Map/Filter), as this might add some non used unnecessary complexity.

But I might not be seeing on the long term, so please correct me if I'm wrong :D

Thanks!

Reply:
The "modifier" function is probably a hack to allow for instance "remove all strings beginning with zzz". The entire function should be replaced with a `StringFilter` function, see https://github.com/kubernetes/kubernetes/pull/103331

Reply:
IMO we MUST NOT introduce a homebrewed type for `[]string`. I think most go progammers would like to use the builtin type as-is. I would not use a such type myself, it obfuscates the code and even experienced go programers would have to check the code/documentation. The alternative is to type-cast when these functions are used which is not very good either. The `Copy` function is BTW not a copy, it is a clone.

And don't re-do well documented and understandable functions like "sort". If you don't want in-place, copy the slice with `[:]`.

A new type replacing the well known `[]string` and these functions will _not_ make the go code more maintainable.

Please keep the built-in types that everybody knows and limit the functions to those that really add some value, not "sort" and "copy". Filter and map are good candidates, but something like
```go
  filteredSlice := utilstrings.FilterSlice(validator, mySlice)
```